### PR TITLE
Kymotracker line connection function

### PR DESCRIPTION
**Pull request by rpauszek**
**Why this PR?**
Pulled out the hit test for manual line connection so the functionality can be reused in other UI

### CODE REVIEW

File: lumicks/pylake/nb_widgets/kymotracker_widgets.py [hunk 1]
```diff
@@ -482,6 +459,76 @@ def show(self, use_widgets, **kwargs):
         self._select_state({"value": "mode", "old": "", "new": "Track lines"})
 
 
+def _get_node_info(lines):
```
Comment (JoepVanlier) [inline lumicks/pylake/nb_widgets/kymotracker_widgets.py:462]: Would be good to add a test for `_get_nearest` now that it's just a free func.

File: lumicks/pylake/nb_widgets/kymotracker_widgets.py [hunk 2]
```diff
@@ -482,6 +459,76 @@ def show(self, use_widgets, **kwargs):
...
+                track.seconds,
+                track.position,
+            )
+        )
+        for j, track in enumerate(lines)
+    ]
+    return np.hstack(nodes).T
+
+
+def _get_nearest(lines, nodes, x, y, scale, cutoff_radius=0.05):
```
Comment (JoepVanlier) [inline lumicks/pylake/nb_widgets/kymotracker_widgets.py:492]: Same here. Would be nice to have a test for this. There's also ample room for optimization here, and it'd be nice to have a test in place if we wanna have a go at optimizing this.

**Review by JoepVanlier — COMMENTED** [review]
Yeah, this is nice. Also nice to get rid of that ugly `nonlocal`. Would like to see a few tests added for the new free funcs, but otherwise it looks nice. :)

File: lumicks/pylake/nb_widgets/kymotracker_widgets.py [hunk 3]
```diff
@@ -140,49 +140,25 @@ def set_xlim(drag_event):
 
         MouseDragCallback(self._axes, 1, set_xlim)
 
-    def _get_scale(self):
-        """Get scaling of the image axes"""
-        return tuple(lims[1] - lims[0] for lims in (self._axes.get_xlim(), self._axes.get_ylim()))
-
     def _connect_line_callback(self):
         canvas = self._axes.figure.canvas
+        scale = tuple(lims[1] - lims[0] for lims in (self._axes.get_xlim(), self._axes.get_ylim()))
```
Comment (mikhas) [inline lumicks/pylake/nb_widgets/kymotracker_widgets.py:145]: Perhaps rename `scale` to `visible_range`?

File: lumicks/pylake/nb_widgets/kymotracker_widgets.py [hunk 4]
```diff
@@ -140,49 +140,25 @@ def set_xlim(drag_event):
...
             if len(self.lines) == 0:
                 return
 
-            nodes = get_node_info()
-            distance, idx = get_nearest(event.x, event.y)
-
-            if distance < cutoff_radius:
-                line_index, *line_info = nodes[idx]
-                clicked_line_info = [self.lines[int(line_index)], *line_info]
+            nodes = _get_node_info(self.lines)
```
Comment (mikhas) [inline lumicks/pylake/nb_widgets/kymotracker_widgets.py:156]: It's not a problem that I now have to call `_get_node_info()` myself, but it does leak implementation details that we don't really need, thus binding you a bit to this admittedly still private API.

All _I_ need, for the `_merge_lines()` to work, is the position of the datapoint inside a `KymoLine`. Maybe we can rethink this part a bit.

File: lumicks/pylake/nb_widgets/kymotracker_widgets.py (hunk 2 above)
Comment (mikhas) [inline lumicks/pylake/nb_widgets/kymotracker_widgets.py:492]: Tested it, works great!

**Review by mikhas — COMMENTED** [review]
Agree with Joep that the tests need to change as well, even though one could argue that the existing integration test with the fake mouse events still covers this. Not too thrilled about the node info API, however.

File: lumicks/pylake/nb_widgets/kymotracker_widgets.py (hunk 4 above)
Comment (rpauszek) [reply lumicks/pylake/nb_widgets/kymotracker_widgets.py]: I agree, the API was a bit weird. basically I left it as it already was, but just pulled out into a separate function. I've updated it now to be a bit more straightforward I think. `_get_node_info()` is back to being truly internal, and `_get_nearest()` returns a `dict` with the relevant info if a node was clicked, else empty.

See last 3 commits for proposed changes, these'll be squashed on merge

**Comment by rpauszek** [comment]
not needed anymore

**Pull request closed without merging**